<div align="center">
    <img src="artic-logo.png?raw=true?" alt="artic-logo" width="250">
    <h1>ARTIC-VIDRL</h1>
    <h3>A fork of the ARTIC pipeline for the Victorian Infectious Diseases Reference Laboratory </h3>
</div>

---

Wrapper implementing a limited ARTIC pipeline (`Medaka`) for non-standard viral amplicon schemes. It expects demultiplexed and adapter trimmed input files (as for example generated by latest versions of `Guppy`).

* `Nextflow` implementation of the `Medaka` workflow 
* `Nanoq` for gathering of compressed and demultiplexed files
* `Longshot` update to functional version (`v0.4.5`)

Validated primer schemes from our work to be published in this repository.

## Dependencies

* `Nextflow >= v21.10`
* `conda` | `mamba` | `micromamba`

## Input formats

Primer scheme directory:

```
/path/to/scheme/dir/{scheme_name}/V{scheme_version}
```

Reference and scheme file **must** conform to the scheme name and extensions:

```
/path/to/scheme/dir/{scheme_name}/V{scheme_version}/{scheme_name}.reference.fasta
/path/to/scheme/dir/{scheme_name}/V{scheme_version}/{scheme_name}.scheme.bed
```

Input read files can be compressed (`gz`, `bz2`, `lzma`).

## Usage

```
nextflow run esteinig/artic-vidrl --help
```

Run the workflow directly from the repository using the `conda` profile option, which will cache a `conda` environment in the working directory (once). If installed, `mamba` or `micromamba` can be activated to replace the slower `conda` installation:

```
nextflow run esteinig/artic-vidrl -profile conda (--mamba true | --micromamba true) <args>
```

### Single sample

A single sample can be gathered from one or multiple files using a wildcard pattern (e.g. from a barcode directory as output by `Guppy`). Note the required quotation marks around the pattern, so that the shell does not intepret the wildcard. An output sample identifier (`--fastq_id`) must be provided.

```
nextflow run esteinig/artic-vidrl -profile conda \
    --scheme_dir EBOV/V1 \
    --medaka_model "r941_min_high_g360" \
    --fastq_gather "fastq/*.fastq.gz" \
    --fastq_id "test" \
    --outdir test_single
```

### Multiple samples

Multiple samples in subdirectories (e.g. demultiplexed barcode directories) can be run by specifying the parent directory and the extension of files to gather in each subdirectory:

```
nextflow run esteinig/artic-vidrl -profile conda \
    --scheme_dir EBOV/V1 \
    --medaka_model "r941_min_high_g360" \
    --fastq_dir read_dir \
    --fastq_ext ".fastq.gz" \
    --outdir test_multiple 
```

The input directory structure would look like this:

```
read_dir/{subdir}/{read_files}.fastq.gz
```

Output identifiers for each sample is the names of the subdirectory.

### Barcodes

When running on demultiplexed barcode outputs, not all barcode subdirectories may be relevant. You can specify a dashed range or a comma-delimited list of barcode numbers (`int`-format: 1, 2, ... 10, 11) that subsets the subdirectories:

```
--barcodes "1-21"
```

or

```
--barcodes "1,2,5,7,21"
```

This will look for trailing barcode numbers (0-padded, as in the standard output for barcodes: 01, 02, ... 10, 11) at the end of the subdirectory names. In this case the input directory structure would look like this:

```
read_dir/{barcode_dir}{barcode_number}/{read_files}.fastq.gz
```

e.g.

```
read_dir/barcode03/{read_files}.fastq.gz
```

### Required arguments

Configuration:

* `--scheme_dir`: the path to the scheme directory including the version subdirectory
* `--medaka_model`: the `Medaka` model to use in the workflow

Input:

* `--fastq_gather`: a quoted wildcard expression to collect read files for a single sample 
* `--fastq_id`: the output identifier for the sample when using `--fastq_gather`

OR

* `--fastq_dir`: the path to a directory with subdirectories for multiple samples
* `--fastq_ext`: the file extension of read files in the subdirectories

#### Scheme directory

Primer scheme directory provided to `Nextflow` must include the version directory:

```
--scheme_dir {scheme_name}/V{scheme_version}
```

Paths to the scheme directory can be relative or absolute.

```
--scheme_dir ${PWD}/{scheme_name}/V{scheme_version}
```

### Local environment

You can run the pipeline in an active local environment but this must be within an `artic` environment with an updated version of `Longshot`. You can use the `environment.yml` file in this repository to install the updated dependencies:

```
git clone https://github.com/esteinig/artic-vidrl
conda env create -f artic-vidrl/environment.yml
conda activate artic-vidrl
```

Default is the `local` executor in the workflow, run without the profile option:

```
nextflow run esteinig/artic-vidrl \
    --scheme_dir EBOV/V1 \
    --medaka_model "r941_min_high_g360" \ 
    --fastq_gather "fastq/*.fastq.gz" \
    --fastq_id "test"
```

Alternatively point the workflow directly at the path of the local `conda` environment, e.g.

```
nextflow run esteinig/artic-vidrl \
    --conda_env /data/opt/conda/envs/artic-vidrl \
    --scheme_dir EBOV/V1 \
    --medaka_model "r941_min_high_g360" \ 
    --fastq_gather "fastq/*.fastq.gz" \
    --fastq_id "test"
```


## Change log

- Added final alignment optional flag to `artic_minion.py`
- Version update: added a wheel file for `artic-v1.2.2` including a `pip` directive in `environment.yml` (this fork) so that we can update `longshot=0.4.5` in `environment.yml` and still install `artic` within the conda handler of Nextflow. it was not possible to use a reduced environment file that specified both `artic=1.2.2` and `longshot=0.4.5`, because of conflict with the `longshot=0.4.1` requirement in `artic=1.2.2`. Also added `nanoq=0.9.0` as dependency for a simple (and more flexible) replacement of `artic gather`.
